llvm/lib/Target/AIE/AIEPostSelectOptimize.cpp

```diff
@@ -438,6 +450,161 @@ bool duplicateAdressingRegs(MachineBasicBlock &MBB, MachineRegisterInfo &MRI) {
   return tryToDuplicateLoadUse(LoadUses, NonLoadUses, MRI, TII);
 }
 
+/// This optimization tries to propagate GlobalValues as MMO of load operations
+/// when they are missing.
+bool findGlobalValues(const MachineInstr *MI,
+                      SmallPtrSet<const Value *, 4> &GVSet,
+                      SmallPtrSet<const MachineInstr *, 4> &VisitedInstrs,
+                      MachineRegisterInfo &MRI, unsigned SearchLimit) {
+
+  // Exhausted the search.
+  if (VisitedInstrs.size() == SearchLimit)
+    return false;
+
+  // Skip copies.
+  while (MI->isCopy() && MI->getOperand(1).getReg().isVirtual())
+    MI = MRI.getVRegDef(MI->getOperand(1).getReg());
+
+  // Loaded values cannot be tracked as Globals, but they can
+  // be part of a final address calculation.
+  if (MI->mayLoad())
+    return true;
+
+  if (VisitedInstrs.find(MI) != VisitedInstrs.end())
+    return true;
+
+  VisitedInstrs.insert(MI);
+  bool Success = true;
+  for (auto MO : MI->uses()) {
+    if (MO.isGlobal()) {
+      GVSet.insert(MO.getGlobal());
+    } else if (MO.isReg() && MO.getReg().isVirtual()) {
+      if (MO.getReg().isVirtual()) {
+        const Register Reg = MO.getReg();
+        if (const MachineInstr *RegDef = MRI.getVRegDef(Reg))
+          Success &=
+              findGlobalValues(RegDef, GVSet, VisitedInstrs, MRI, SearchLimit);
+      } else {
+        // Physical registers can be anything (parameters).
+        return false;
+      }
+    }
+  }
+  return Success;
+}
+
+using Opcodes = AIEBaseInstrInfo::AbstractVecOp::AbstractOpcode;
+using AbstractVecOp = AIEBaseInstrInfo::AbstractVecOp;
+using OptionalVecOp = std::optional<AbstractVecOp>;
+
+/// This optimization tries to propagate GlobalValues as MMO of load operations
+/// when they are missing. This specific implementation considers only the
+/// following pattern:
+/// VADD (VSEL (VBCST(GlobalValue1), VBCST(GlobalValue2), Mask) Value)
+bool findBroadcastedGlobalValues(const MachineInstr *MI,
+                                 SmallPtrSet<const Value *, 4> &GVSet,
+                                 MachineRegisterInfo &MRI) {
+
+  const TargetSubtargetInfo &ST = MI->getParent()->getParent()->getSubtarget();
+  const AIEBaseInstrInfo *TII =
+      static_cast<const AIEBaseInstrInfo *>(ST.getInstrInfo());
+
+  // Skip subvector copy.
+  if (MI->isCopy()) {
```

Comment:
Nit: I guess there could be multiple copies. Isn't there a helper function somewhere to look through copies?